macOS/unwindmacos.cpp

```diff
@@ -288,64 +470,110 @@ bool UnwindGeneratorMacOS::HandleBreakpoint(ModuleInfo* module, void *address) {
   return true;
 }
 
-// the idea is that this produces a minimal valid FDE
-// but since __register_frame doesn't return a value
-// it's not possible to test if the registration actually succeeds
-size_t UnwindGeneratorMacOS::WriteTestFde(ModuleInfo *module) {
-  // need to write a CIE first
-  // see CFI_Parser<A>::parseCIE
-  unsigned char test_cie[] =
-  {
-    0x0d, 0x00, 0x00, 0x00, // CIE length
-    0x00, 0x00, 0x00, 0x00, // CIE id, must be zero
-    0x03, // version
-    0x00, // empty augmentation string
-    0x00, // code aligment factor, LEB128 encoded
-    0x00, // data aligment factor, LEB128 encoded
-    0x00 // return address register, LEB128 encoded
-  };
-  tinyinst_.WriteCode(module, test_cie, sizeof(test_cie));
-  
+void UnwindGeneratorMacOS::WriteDWARFInstructionsRBP(ByteStream *fde, uint32_t encoding) {
+  fde->PutBytes<uint8_t>(DW_CFA_advance_loc | 1);
+
+  fde->PutBytes<uint8_t>(DW_CFA_def_cfa_offset);
+  fde->PutEncodedULEB128Bytes(16);
+
+  fde->PutBytes<uint8_t>(DW_CFA_offset | 6);
+  fde->PutEncodedULEB128Bytes(-16);
+
+  fde->PutBytes<uint8_t>(DW_CFA_advance_loc | 3);
+
+  fde->PutBytes<uint8_t>(DW_CFA_def_cfa_register);
+  fde->PutEncodedULEB128Bytes(6);
+
+  uint32_t saved_registers_cfa_offset = -16 - 8 * EXTRACT_BITS(encoding, UNWIND_X86_64_RBP_FRAME_OFFSET);
+  uint32_t saved_registers_locations = EXTRACT_BITS(encoding, UNWIND_X86_64_RBP_FRAME_REGISTERS);
+
+  for (int i = 0; i < 5; i++) {
+    switch (saved_registers_locations & 0x7) {
+      case UNWIND_X86_64_REG_NONE:
+        break;
+      case UNWIND_X86_64_REG_RBX:
+        fde->PutBytes<uint8_t>(DW_CFA_offset | 3);
+        break;
+      case UNWIND_X86_64_REG_R12:
+        fde->PutBytes<uint8_t>(DW_CFA_offset | 12);
+        break;
+      case UNWIND_X86_64_REG_R13:
+        fde->PutBytes<uint8_t>(DW_CFA_offset | 13);
+        break;
+      case UNWIND_X86_64_REG_R14:
+        fde->PutBytes<uint8_t>(DW_CFA_offset | 14);
+        break;
+      case UNWIND_X86_64_REG_R15:
+        fde->PutBytes<uint8_t>(DW_CFA_offset | 15);
+        break;
+    }
+
+    if ((saved_registers_locations & 0x7) != UNWIND_X86_64_REG_NONE) {
+      fde->PutEncodedULEB128Bytes(saved_registers_cfa_offset);
+    }
+
+    saved_registers_cfa_offset += 8;
+    saved_registers_locations >>= 3;
+  }
+
+//  fde->PutBytes<uint8_t>(DW_CFA_def_cfa);
+//  fde->PutEncodedULEB128Bytes(7);
+//  fde->PutEncodedULEB128Bytes(8);
+}
+
+void UnwindGeneratorMacOS::WriteDWARFInstructions(ByteStream *fde, uint32_t encoding) {
+  uint32_t mode = encoding & UNWIND_X86_64_MODE_MASK;
+  switch(mode) {
+    case UNWIND_X86_64_MODE_RBP_FRAME:
+      WriteDWARFInstructionsRBP(fde, encoding);
+      break;
+    default:
+      WARN("Unsupported encoding");
```

Comment:
Perhaps could be useful to print out the value of the unsupported encoding here?